File: platform/how-to-get-started/2.fetch_httpx_gemini_openai_protocols/2.gemini-openai-tool-calls.py
#!/usr/bin/env python3
"""
Example: OpenAI ↔ Gemini Tool Call Normalization

This example demonstrates:
1. Converting OpenAI tool definitions to Gemini function declarations
2. Converting OpenAI tool_calls to Gemini functionCall parts
3. Converting Gemini function calls back to OpenAI tool_calls
4. Handling tool_choice / toolConfig translation

The translation handles:
- Tool definition schema translation
- Tool call ID generation
- Arguments serialization (JSON string ↔ object)
"""
import os
import sys
import json
import asyncio
from pathlib import Path

# Add packages to path
ROOT_DIR = Path(__file__).parent.parent.parent
sys.path.insert(0, str(ROOT_DIR / "packages_py"))
sys.path.insert(0, str(ROOT_DIR / "packages_py/fetch_httpx_gemini_openai_constant"))

from fetch_httpx import AsyncClient, Timeout
from fetch_httpx_gemini_openai_constant import (
    GEMINI_ORIGIN,
    GEMINI_CONNECT_TIMEOUT_S,
    GEMINI_READ_TIMEOUT_THINKING_S,
)
from fetch_httpx_gemini_openai_protocols import (
    openai_tools_to_gemini,
    gemini_tools_to_openai,
    openai_tool_calls_to_gemini,
    gemini_function_calls_to_openai,
    openai_tool_choice_to_gemini,
    gemini_tool_config_to_openai,
    translate_openai_request_to_gemini,
    translate_gemini_response_to_openai,
    generate_tool_call_id,
)

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
if not GEMINI_API_KEY:
    print("ERROR: GEMINI_API_KEY environment variable not set")
    print("Set it with: export GEMINI_API_KEY=your_api_key")
    sys.exit(1)


async def main():
    # =========================================================================
    # Example 1: Tool Definition Translation
    # =========================================================================

    print("=" * 60)
    print("Example 1: OpenAI → Gemini Tool Definition Translation")
    print("=" * 60)

    # OpenAI-style tool definitions
    openai_tools = [
        {
            "type": "function",
            "function": {
                "name": "get_weather",
                "description": "Get the current weather for a location",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {
                            "type": "string",
                            "description": "The city and state, e.g., San Francisco, CA",
                        },
                        "unit": {
                            "type": "string",
                            "enum": ["celsius", "fahrenheit"],
                            "description": "Temperature unit",
                        },
                    },
                    "required": ["location"],
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "search_web",
                "description": "Search the web for information",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {
                            "type": "string",
                            "description": "The search query",
                        },
                        "max_results": {
                            "type": "number",
                            "description": "Maximum number of results",
                        },
                    },
                    "required": ["query"],
                },
            },
        },
    ]

    print("\nOpenAI Tools:")
    print(json.dumps(openai_tools, indent=2))

    # Translate to Gemini format
    gemini_tools_result = openai_tools_to_gemini(openai_tools)

    print("\nGemini Tools:")
    print(json.dumps(gemini_tools_result.data, indent=2))

    # =========================================================================
    # Example 2: Tool Call Translation (OpenAI → Gemini)
    # =========================================================================

    print("\n" + "=" * 60)
    print("Example 2: OpenAI → Gemini Tool Call Translation")
    print("=" * 60)

    # OpenAI-style tool calls (from assistant response)
    openai_tool_calls = [
        {
            "id": "call_abc123",
            "type": "function",
            "function": {
                "name": "get_weather",
                "arguments": json.dumps({"location": "San Francisco, CA", "unit": "fahrenheit"}),
            },
        },
        {
            "id": "call_def456",
            "type": "function",
            "function": {
                "name": "search_web",
                "arguments": json.dumps({"query": "best restaurants SF", "max_results": 5}),
            },
        },
    ]

    print("\nOpenAI Tool Calls:")
    print(json.dumps(openai_tool_calls, indent=2))

    # Translate to Gemini format
    gemini_parts_result = openai_tool_calls_to_gemini(openai_tool_calls)

    print("\nGemini Function Call Parts:")
    print(json.dumps(gemini_parts_result.data, indent=2))

    # =========================================================================
    # Example 3: Tool Call Translation (Gemini → OpenAI)
    # =========================================================================

    print("\n" + "=" * 60)
    print("Example 3: Gemini → OpenAI Tool Call Translation")
    print("=" * 60)

    # Gemini-style function call parts
    gemini_function_call_parts = [
        {
            "functionCall": {
                "name": "get_weather",
                "args": {"location": "New York, NY", "unit": "celsius"},
            },
        },
        {
            "functionCall": {
                "name": "search_web",
                "args": {"query": "weather forecast NYC"},
            },
        },
    ]

    print("\nGemini Function Call Parts:")
    print(json.dumps(gemini_function_call_parts, indent=2))

    # Translate to OpenAI format
    openai_tool_calls_result = gemini_function_calls_to_openai(gemini_function_call_parts)

    print("\nOpenAI Tool Calls:")
    print(json.dumps(openai_tool_calls_result.data, indent=2))

    # =========================================================================
    # Example 4: Tool Choice Translation
    # =========================================================================

    print("\n" + "=" * 60)
    print("Example 4: Tool Choice / Tool Config Translation")
    print("=" * 60)

    tool_choices = [
        "none",
        "auto",
        "required",
        {"type": "function", "function": {"name": "get_weather"}},
    ]

    for choice in tool_choices:
        gemini_config = openai_tool_choice_to_gemini(choice)
        print(f"\nOpenAI tool_choice: {json.dumps(choice)}")
        print(f"Gemini toolConfig: {json.dumps(gemini_config)}")

        # Translate back
        back_to_openai = gemini_tool_config_to_openai(gemini_config)
        print(f"Back to OpenAI: {json.dumps(back_to_openai)}")

    # =========================================================================
    # Example 5: Live API Call with Tool Calls
    # =========================================================================

    print("\n" + "=" * 60)
    print("Example 5: Live API Call with Tool Calls")
    print("=" * 60)

    # OpenAI-style request with tools
    openai_request = {
        "model": "gemini-2.0-flash",
        "messages": [
            {
                "role": "system",
                "content": "You are a helpful assistant. Use the provided tools when appropriate.",
            },
            {"role": "user", "content": "What is the weather in Tokyo?"},
        ],
        "tools": [
            {
                "type": "function",
                "function": {
                    "name": "get_weather",
                    "description": "Get the current weather for a location",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "location": {"type": "string", "description": "City name"},
                            "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
                        },
                        "required": ["location"],
                    },
                },
            },
        ],
        "tool_choice": "auto",
    }

    print("\nOpenAI-style Request:")
    print(json.dumps(openai_request, indent=2))

    # Translate to Gemini
    translated_request = translate_openai_request_to_gemini(openai_request)
    print("\nTranslated to Gemini Request:")
    print(json.dumps(translated_request.data, indent=2))

    # Make API call to Gemini's native endpoint
    GEMINI_URL = f"{GEMINI_ORIGIN}/v1beta/models/gemini-2.0-flash:generateContent"

    async with AsyncClient(timeout=Timeout(connect=GEMINI_CONNECT_TIMEOUT_S, read=GEMINI_READ_TIMEOUT_THINKING_S)) as client:
        try:
            response = await client.post(
                GEMINI_URL,
                headers={
                    "Content-Type": "application/json",
                    "x-goog-api-key": GEMINI_API_KEY,
                },
                json=translated_request.data,
            )

            if response.is_success:
                gemini_response = response.json()
                print("\nGemini Native Response:")
                print(json.dumps(gemini_response, indent=2))

                # Translate response back to OpenAI format
                openai_response = translate_gemini_response_to_openai(
                    gemini_response, model="gemini-2.0-flash"
                )
                print("\nTranslated to OpenAI Response:")
                print(json.dumps(openai_response.data, indent=2))

                # Check for tool calls
                choice = openai_response.data.get("choices", [{}])[0]
                tool_calls = choice.get("message", {}).get("tool_calls")
                if tool_calls:
                    print("\nTool Calls Detected:")
                    for tc in tool_calls:
                        print(f"  - {tc['function']['name']}({tc['function']['arguments']})")
                elif choice.get("message", {}).get("content"):
                    print("\nText Response:", choice["message"]["content"])
            else:
                print(f"\nAPI Error: {response.status_code} {response.text}")

        except Exception as e:
            print(f"\nRequest Error: {e}")

    # =========================================================================
    # Example 6: Gemini → OpenAI Tool Definition Translation
    # =========================================================================

    print("\n" + "=" * 60)
    print("Example 6: Gemini → OpenAI Tool Definition Translation")
    print("=" * 60)

    # Gemini-style tools
    gemini_tools = [
        {
            "functionDeclarations": [
                {
                    "name": "calculate",
                    "description": "Perform a mathematical calculation",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "expression": {"type": "string", "description": "Math expression to evaluate"},
                        },
                        "required": ["expression"],
                    },
                },
                {
                    "name": "translate",
                    "description": "Translate text to another language",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "text": {"type": "string", "description": "Text to translate"},
                            "target_language": {"type": "string", "description": "Target language code"},
                        },
                        "required": ["text", "target_language"],
                    },
                },
            ],
        },
    ]

    print("\nGemini Tools:")
    print(json.dumps(gemini_tools, indent=2))

    # Translate to OpenAI format
    openai_tools_result = gemini_tools_to_openai(gemini_tools)

    print("\nOpenAI Tools:")
    print(json.dumps(openai_tools_result.data, indent=2))

    print("\n" + "=" * 60)
    print("Tool Call Examples Complete")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())

File: platform/how-to-get-started/2.fetch_httpx_gemini_openai_protocols/2.gemini-openai-tool-calls.mjs
#!/usr/bin/env node
/**
 * Example: OpenAI ↔ Gemini Tool Call Normalization
 *
 * This example demonstrates:
 * 1. Converting OpenAI tool definitions to Gemini function declarations
 * 2. Converting OpenAI tool_calls to Gemini functionCall parts
 * 3. Converting Gemini function calls back to OpenAI tool_calls
 * 4. Handling tool_choice / toolConfig translation
 *
 * The translation handles:
 * - Tool definition schema translation
 * - Tool call ID generation
 * - Arguments serialization (JSON string ↔ object)
 */
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const ROOT_DIR = join(__dirname, '../..');

// Import protocol translation
const {
    openAIToolsToGemini,
    geminiToolsToOpenAI,
    openAIToolCallsToGemini,
    geminiFunctionCallsToOpenAI,
    openAIToolChoiceToGemini,
    geminiToolConfigToOpenAI,
    translateOpenAIRequestToGemini,
    translateGeminiResponseToOpenAI,
    generateToolCallId,
} = await import(
    join(ROOT_DIR, 'packages_mjs/fetch_undici_gemini_openai_protocols/dist/index.js')
);

// Import fetch client
const { AsyncClient, Timeout } = await import(
    join(ROOT_DIR, 'packages_mjs/fetch_undici/dist/index.js')
);

// Import Gemini constants
const {
    GEMINI_ORIGIN,
    GEMINI_CONNECT_TIMEOUT_MS,
    GEMINI_READ_TIMEOUT_THINKING_MS,
} = await import(
    join(ROOT_DIR, 'packages_mjs/fetch_undici_gemini_openai_constant/dist/index.js')
);

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
if (!GEMINI_API_KEY) {
    console.error('ERROR: GEMINI_API_KEY environment variable not set');
    console.error('Set it with: export GEMINI_API_KEY=your_api_key');
    process.exit(1);
}

// =============================================================================
// Example 1: Tool Definition Translation
// =============================================================================

console.log('='.repeat(60));
console.log('Example 1: OpenAI → Gemini Tool Definition Translation');
console.log('='.repeat(60));

// OpenAI-style tool definitions
const openAITools = [
    {
        type: 'function',
        function: {
            name: 'get_weather',
            description: 'Get the current weather for a location',
            parameters: {
                type: 'object',
                properties: {
                    location: {
                        type: 'string',
                        description: 'The city and state, e.g., San Francisco, CA',
                    },
                    unit: {
                        type: 'string',
                        enum: ['celsius', 'fahrenheit'],
                        description: 'Temperature unit',
                    },
                },
                required: ['location'],
            },
        },
    },
    {
        type: 'function',
        function: {
            name: 'search_web',
            description: 'Search the web for information',
            parameters: {
                type: 'object',
                properties: {
                    query: {
                        type: 'string',
                        description: 'The search query',
                    },
                    max_results: {
                        type: 'number',
                        description: 'Maximum number of results',
                    },
                },
                required: ['query'],
            },
        },
    },
];

console.log('\nOpenAI Tools:');
console.log(JSON.stringify(openAITools, null, 2));

// Translate to Gemini format
const geminiToolsResult = openAIToolsToGemini(openAITools);

console.log('\nGemini Tools:');
console.log(JSON.stringify(geminiToolsResult.data, null, 2));

// =============================================================================
// Example 2: Tool Call Translation (OpenAI → Gemini)
// =============================================================================

console.log('\n' + '='.repeat(60));
console.log('Example 2: OpenAI → Gemini Tool Call Translation');
console.log('='.repeat(60));

// OpenAI-style tool calls (from assistant response)
const openAIToolCalls = [
    {
        id: 'call_abc123',
        type: 'function',
        function: {
            name: 'get_weather',
            arguments: JSON.stringify({ location: 'San Francisco, CA', unit: 'fahrenheit' }),
        },
    },
    {
        id: 'call_def456',
        type: 'function',
        function: {
            name: 'search_web',
            arguments: JSON.stringify({ query: 'best restaurants SF', max_results: 5 }),
        },
    },
];

console.log('\nOpenAI Tool Calls:');
console.log(JSON.stringify(openAIToolCalls, null, 2));

// Translate to Gemini format
const geminiPartsResult = openAIToolCallsToGemini(openAIToolCalls);

console.log('\nGemini Function Call Parts:');
console.log(JSON.stringify(geminiPartsResult.data, null, 2));

// =============================================================================
// Example 3: Tool Call Translation (Gemini → OpenAI)
// =============================================================================

console.log('\n' + '='.repeat(60));
console.log('Example 3: Gemini → OpenAI Tool Call Translation');
console.log('='.repeat(60));

// Gemini-style function call parts
const geminiFunctionCallParts = [
    {
        functionCall: {
            name: 'get_weather',
            args: { location: 'New York, NY', unit: 'celsius' },
        },
    },
    {
        functionCall: {
            name: 'search_web',
            args: { query: 'weather forecast NYC' },
        },
    },
];

console.log('\nGemini Function Call Parts:');
console.log(JSON.stringify(geminiFunctionCallParts, null, 2));

// Translate to OpenAI format
const openAIToolCallsResult = geminiFunctionCallsToOpenAI(geminiFunctionCallParts);

console.log('\nOpenAI Tool Calls:');
console.log(JSON.stringify(openAIToolCallsResult.data, null, 2));

// =============================================================================
// Example 4: Tool Choice Translation
// =============================================================================

console.log('\n' + '='.repeat(60));
console.log('Example 4: Tool Choice / Tool Config Translation');
console.log('='.repeat(60));

const toolChoices = ['none', 'auto', 'required', { type: 'function', function: { name: 'get_weather' } }];

for (const choice of toolChoices) {
    const geminiConfig = openAIToolChoiceToGemini(choice);
    console.log(`\nOpenAI tool_choice: ${JSON.stringify(choice)}`);
    console.log(`Gemini toolConfig: ${JSON.stringify(geminiConfig)}`);

    // Translate back
    const backToOpenAI = geminiToolConfigToOpenAI(geminiConfig);
    console.log(`Back to OpenAI: ${JSON.stringify(backToOpenAI)}`);
}

// =============================================================================
// Example 5: Live API Call with Tool Calls
// =============================================================================

console.log('\n' + '='.repeat(60));
console.log('Example 5: Live API Call with Tool Calls');
console.log('='.repeat(60));

// OpenAI-style request with tools
const openAIRequest = {
    model: 'gemini-2.0-flash',
    messages: [
        { role: 'system', content: 'You are a helpful assistant. Use the provided tools when appropriate.' },
        { role: 'user', content: 'What is the weather in Tokyo?' },
    ],
    tools: [
        {
            type: 'function',
            function: {
                name: 'get_weather',
                description: 'Get the current weather for a location',
                parameters: {
                    type: 'object',
                    properties: {
                        location: { type: 'string', description: 'City name' },
                        unit: { type: 'string', enum: ['celsius', 'fahrenheit'] },
                    },
                    required: ['location'],
                },
            },
        },
    ],
    tool_choice: 'auto',
};

console.log('\nOpenAI-style Request:');
console.log(JSON.stringify(openAIRequest, null, 2));

// Translate to Gemini
const translatedRequest = translateOpenAIRequestToGemini(openAIRequest);
console.log('\nTranslated to Gemini Request:');
console.log(JSON.stringify(translatedRequest.data, null, 2));

// Make API call to Gemini's native endpoint
const GEMINI_URL = `${GEMINI_ORIGIN}/v1beta/models/gemini-2.0-flash:generateContent`;

const client = new AsyncClient({
    timeout: new Timeout({ connect: GEMINI_CONNECT_TIMEOUT_MS, read: GEMINI_READ_TIMEOUT_THINKING_MS }),
});

try {
    const response = await client.post(GEMINI_URL, {
        headers: {
            'Content-Type': 'application/json',
            'x-goog-api-key': GEMINI_API_KEY,
        },
        json: translatedRequest.data,
    });

    if (response.isSuccess) {
        const geminiResponse = await response.json();
        console.log('\nGemini Native Response:');
        console.log(JSON.stringify(geminiResponse, null, 2));

        // Translate response back to OpenAI format
        const openAIResponse = translateGeminiResponseToOpenAI(geminiResponse, 'gemini-2.0-flash');
        console.log('\nTranslated to OpenAI Response:');
        console.log(JSON.stringify(openAIResponse.data, null, 2));

        // Check for tool calls
        const choice = openAIResponse.data.choices?.[0];
        if (choice?.message?.tool_calls) {
            console.log('\nTool Calls Detected:');
            for (const toolCall of choice.message.tool_calls) {
                console.log(`  - ${toolCall.function.name}(${toolCall.function.arguments})`);
            }
        } else if (choice?.message?.content) {
            console.log('\nText Response:', choice.message.content);
        }
    } else {
        const errorText = await response.text();
        console.error('\nAPI Error:', response.statusCode, errorText);
    }
} catch (error) {
    console.error('\nRequest Error:', error.message);
} finally {
    await client.close();
}

// =============================================================================
// Example 6: Gemini → OpenAI Tool Definition Translation
// =============================================================================

console.log('\n' + '='.repeat(60));
console.log('Example 6: Gemini → OpenAI Tool Definition Translation');
console.log('='.repeat(60));

// Gemini-style tools
const geminiTools = [
    {
        functionDeclarations: [
            {
                name: 'calculate',
                description: 'Perform a mathematical calculation',
                parameters: {
                    type: 'object',
                    properties: {
                        expression: { type: 'string', description: 'Math expression to evaluate' },
                    },
                    required: ['expression'],
                },
            },
            {
                name: 'translate',
                description: 'Translate text to another language',
                parameters: {
                    type: 'object',
                    properties: {
                        text: { type: 'string', description: 'Text to translate' },
                        target_language: { type: 'string', description: 'Target language code' },
                    },
                    required: ['text', 'target_language'],
                },
            },
        ],
    },
];

console.log('\nGemini Tools:');
console.log(JSON.stringify(geminiTools, null, 2));

// Translate to OpenAI format
const openAIToolsResult = geminiToolsToOpenAI(geminiTools);

console.log('\nOpenAI Tools:');
console.log(JSON.stringify(openAIToolsResult.data, null, 2));

console.log('\n' + '='.repeat(60));
console.log('Tool Call Examples Complete');
console.log('='.repeat(60));
